maven-release-manager/src/main/java/org/apache/maven/shared/release/util/CiFriendlyVersion.java

```diff
@@ -0,0 +1,96 @@
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.maven.shared.release.util;
+
+import java.util.Arrays;
+import java.util.List;
+import java.util.Properties;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+import org.apache.maven.artifact.ArtifactUtils;
+
+public class CiFriendlyVersion {
+
+    /**
+     * Regular expression pattern matching Maven expressions (i.e. references to Maven properties).
+     * The first group selects the property name the expression refers to.
+     */
+    private static final Pattern EXPRESSION_PATTERN = Pattern.compile("\\$\\{(.+?)\\}");
+
+    /**
+     * All Maven properties allowed to be referenced in parent versions via expressions
+     * @see <a href="https://maven.apache.org/maven-ci-friendly.html">CI-Friendly Versions</a>
+     */
+    private static final String REVISION = "revision";
+
+    private static final String SHA_1 = "sha1";
+    private static final String CHANGELIST = "changelist";
+    private static final List<String> CI_FRIENDLY_PROPERTIES = Arrays.asList(REVISION, SHA_1, CHANGELIST);
+
+    private static final String SNAPSHOT = "-SNAPSHOT";
+
+    private CiFriendlyVersion() {}
+
+    /**
+     * Extracts the Maven property name from a given expression.
+     * @param expression the expression
+     * @return either {@code null} if value is no expression otherwise the property referenced in the expression
+     */
+    public static String extractPropertyFromExpression(String expression) {
+        Matcher matcher = EXPRESSION_PATTERN.matcher(expression);
+        if (!matcher.find()) {
+            return null;
+        }
+        return matcher.group(1);
+    }
+
+    public static boolean isCiFriendlyVersion(String version) {
+        return containsCiFriendlyProperties(extractPropertyFromExpression(version));
+    }
+
+    public static boolean containsCiFriendlyProperties(String property) {
+        return CI_FRIENDLY_PROPERTIES.contains(property);
+    }
+
+    public static void rewriteVersionAndProperties(String version, String versionElement, Properties properties) {
+        // try to rewrite property if CI friendly expression is used
+        String ciFriendlyPropertyName = extractPropertyFromExpression(versionElement);
+        if (properties != null) {
+            String sha1 = properties.getProperty(SHA_1, System.getProperty(SHA_1, ""));
```

Comment:
Was `System.getProperty()` used before? We tend *not* to rely on system properties, but only on Maven properties.

@cstamas @slawekjaranowski

Reply:
The `sha1` property for CI is usually passed from a command-line. This class is an extract of everything related to CI Friendly. See, the usage of this method, it's used in `JDomModel` which does not have Maven project reference.
The goal of `CiFriendlyVersion` is a small clean up without a big refactoring.

Reply:
This is a contradiction to https://github.com/apache/maven/blob/5e977034565ffc50c6b073f21fab50fae71705dd/maven-api-impl/src/main/java/org/apache/maven/internal/impl/model/DefaultModelVersionProcessor.java#L51-L62 because they can only come from model or Maven user properties...

Reply:
Done

Reply:
I'll revert it back to `System` those properties do not exist in `MavenProject` because the current maven version for this plugin defined as `<mavenVersion>3.2.5</mavenVersion>` but you've put a link to the changes made 2 months ago.

Reply:
That is not correct. The class has been modified long time ago by me. Check its history. I can find the JIRA, if you want. The release of Maven Release will upgrade to 3.6.3 anyway.

Reply:
Here is the change: https://issues.apache.org/jira/browse/MNG-7658. Model properties do get overwritten with the user properties. It is expected to work properly.

Reply:
But it does not with the current dependencies defined in this project :) Just run the integration test and you will see it.

Then the huge problem is the current dependencies defined in the pom.xml. It's just dependency hell that does not work if everything is set up as it's defined.
First of all, `mavenVersion=3.2.5`, the project even does not compile with `maven-3.2.5` binary, it is immediately failed by `maven-enforce-plugin` and requires `3.6.3` version.
Then if you try to bump `mavenVersion` to 3.6.3 and the binary to the same version, it fails again. 

Taking all this dependency incompatibility issues, please do not refer to the recent changes, they just don't work.

Reply:
Okay, I will update the PR later, it will be nice if you notice me in this thread as soon as you finish maven version upgrade.